Addition of More Pooling Methods

billpsomas opened the pull request:
Hi there!

I am the author of the ICCV 2023 paper titled "Keep It SimPool: Who Said Supervised Transformers Suffer from Attention Deficit?", which focuses on benchmarking pooling techniques for CNNs and Transformers. It also introduces a new, simple, attention-based pooling mechanism with great localization properties.

In this pull request, I have implemented and rigorously tested the following pooling methods:

- Generalized Max Pooling
- LSE Pooling 
- HOW Pooling
- Slot Attention (Pooling)
- SimPool
- ViT Pooling

I believe these additions will be beneficial to the library, offering users cutting-edge options for pooling in their models. These methods have shown promising results in my research and experiments, and I am excited about their potential impact on a wider range of applications.

I am looking forward to your feedback and am happy to make any further adjustments as needed. 
Thank you for considering this contribution to the library.

Cheers :)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/timm/pr_2048). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

billpsomas commented:
Hello Ross and sorry for the late reply,

I've indeed integrated everything into my own timm fork and have made a lot of experiments for my paper. You can find experimental results using ResNet-18 in Figure 3 of my paper (https://arxiv.org/pdf/2309.06891.pdf). I've also tested some of them with ResNet-50 and ViT-S. For me, everything worked well. In Figure 3 you will notice that even more poolings are included. I have also integrated these into my timm fork, but did not add them here. Maybe in another PR. 

Now, about attribution as to where the code came from:

GMP: https://github.com/VChristlein/dgmp/blob/master/dgmp.py
LSE: custom implementation
HOW: https://github.com/gtolias/how/tree/master/how/layers
Slot Attention: modified https://github.com/evelinehong/slot-attention-pytorch/blob/master/model.py
SimPool: custom implementation
ViT: https://github.com/sooftware/speech-transformer/blob/master/speech_transformer/attention.py

I know this is not priority, but it would be nice to have some extra poolings in the library. 

In my case, I modified https://github.com/huggingface/pytorch-image-models/blob/main/timm/layers/adaptive_avgmax_pool.py#L124-L161, so that you can give the pooling of your choice through the pool_type argument. 

Cheers!
